### PR TITLE
CLOUDP-199422: deployments connect --type ATLAS

### PR DESCRIPTION
**Pull request by tibulca**
## Proposed changes
implement `atlas deployments connect --type ATLAS`

_Jira ticket:_ CLOUDP-199422

### CODE REVIEW

File: internal/cli/atlas/deployments/connect.go [hunk 1]
```diff
@@ -17,120 +17,58 @@ package deployments
...
+			return Run(cmd.Context(), opts)
 		},
 	}
 
-	cmd.Flags().StringVar(&opts.connectWith, flag.ConnectWith, "", usage.ConnectWith)
+	cmd.Flags().StringVar(&opts.ConnectWith, flag.ConnectWith, "", usage.ConnectWith)
+	cmd.Flags().StringVar(&opts.ProjectID, flag.ProjectID, "", usage.ProjectID)
+	cmd.Flags().StringVar(&opts.DeploymentType, flag.TypeFlag, "", usage.DeploymentType)
+	cmd.Flags().StringVar(&opts.DBUsername, flag.Username, "", usage.DBUsername)
+	cmd.Flags().StringVar(&opts.DBUserPassword, flag.Password, "", usage.Password)
```
Comment (gssbzn) [inline internal/cli/atlas/deployments/connect.go:63]: I don't see it but does this support providing the password via stdin?

File: internal/cli/atlas/deployments/options/deployment_opts.go [hunk 2]
```diff
@@ -222,9 +228,12 @@ func (opts *DeploymentOpts) PromptDeploymentType() error {
 		return err
 	}
 
-	if !strings.EqualFold(opts.DeploymentType, AtlasCluster) && !strings.EqualFold(opts.DeploymentType, LocalCluster) {
-		return fmt.Errorf("%w: %s", errDeploymentTypeNotImplemented, deploymentTypeDescription[opts.DeploymentType])
-	}
+	return ValidateDeploymentType(opts.DeploymentType)
```
Comment (fmenezes) [inline internal/cli/atlas/deployments/options/deployment_opts.go:231]: [q] is this validation needed given we are using a select?

File: internal/cli/atlas/deployments/connect.go (hunk 1 above)
Comment (tibulca) [reply internal/cli/atlas/deployments/connect.go:63]: yes, replied here https://github.com/mongodb/mongodb-atlas-cli/pull/2320/files#r1338813122

File: internal/cli/atlas/deployments/options/deployment_opts.go (hunk 2 above)
Comment (tibulca) [reply internal/cli/atlas/deployments/options/deployment_opts.go]: great catch, there's no need to validate there, but looking at how this PromptDeploymentType func was used I realized we had few small validation bugs in few other commands: we were prompting and validating the type only if it was not provided as a flag. But the commands was accepting any value for the --type flag.
I fixed this in last commit

File: internal/cli/atlas/deployments/options/connect_opts_atlas.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+func (opts *ConnectToAtlasOpts) promptDBUsername() error {
+	p := &survey.Input{
+		Message: "Username for authenticating to MongoDB deployment",
+	}
+	return telemetry.TrackAskOne(p, &opts.DBUsername)
+}
+
+func (opts *ConnectToAtlasOpts) promptDBUserPassword() error {
+	p := &survey.Password{
```
[the remark this answers is not in the document]
Comment (tibulca) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: thanks for catching this, addressed in last commit and tested with
`echo $DBPWD | cli-dev deployments connect Cluster0 --type atlas --connectWith compass --username user1`

File: docs/atlascli/command/atlas-deployments-connect.txt [hunk 4]
```diff
@@ -55,10 +55,30 @@ Options
      - string
      - false
      - Method for connecting to the deployment. Valid values are mongosh, compass, and skip.
+   * - --connectionStringType
+     - string
+     - false
+     - When set to 'private', retrieves the connection string for the network peering endpoint. This value defaults to "standard".
```
Comment (jwilliams-mongo) [inline docs/atlascli/command/atlas-deployments-connect.txt:61]: I don't see where this line was defined in the code, so leaving this comment here. We're using both single quotes and double quotes here. I think we should settle on double quotes.

also a question: are `private' and `standard` the only accepted values?
```suggestion
     - When set to "private", retrieves the connection string for the network peering endpoint. This value defaults to "standard".
```

File: internal/cli/atlas/deployments/options/connect_opts_atlas.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmenezes) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: I wonder if we shouldn't have a flag here? like `--passwordStdIn` or something?

I have a concern that users might end up sending data that are not passwords or vice versa.

Also it helps with discoverability

[the remark this answers is not in the document]
Comment (fmenezes) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: also mongosh and compass have no stdin their way of making it secure is to change process title so we don't leak into ps, I wonder if we shouldn't follow the same approach?

[the remark this answers is not in the document]
Comment (fmenezes) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: looks like we don't have a standard library way of doing this but here there are some suggestions https://github.com/golang/go/issues/4053

File: docs/atlascli/command/atlas-deployments-connect.txt (hunk 4 above)
Comment (tibulca) [reply docs/atlascli/command/atlas-deployments-connect.txt:61]: the first half of the message comes from https://github.com/mongodb/mongodb-atlas-cli/blob/a100b707ff7f008bc12fe18cf54a2a31206fa317/internal/usage/usage.go#L51 and is shared with [atlas-clusters-connectionStrings-describe](https://www.mongodb.com/docs/atlas/cli/stable/command/atlas-clusters-connectionStrings-describe/). 
Second half comes from cobra (the CLI lib).

I can make the change and use double quotes ("private") but looking at usage.go there are more messages using single quotes (e.g. 'forever'). Do you think we should update those too?

For the second question, yes, private and standard are the only options for this flag

File: internal/cli/atlas/deployments/options/connect_opts_atlas.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tibulca) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: this approach (with stdin) is used in 8 more commands. If we have strong reasons to update this approach, we should probably do it in a separate task.

[the remark this answers is not in the document]
Comment (fmenezes) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: Let's keep on the pattern we have, I do think that might be value on checking what is the best approach moving forward.

I'm a little hesitant that our pattern does not have a flag to indicate when to read from stdIn but that is for another time.

[the remark this answers is not in the document]
Comment (fmenezes) [reply internal/cli/atlas/deployments/options/connect_opts_atlas.go:111]: a classic example is `passwd --stdin <username>` where to provide the password via stdin you have to pass the flag

File: docs/atlascli/command/atlas-deployments-connect.txt (hunk 4 above)
Comment (tibulca) [reply docs/atlascli/command/atlas-deployments-connect.txt:61]: Merging this PR and will create a separate PR based on the suggestions that will result from this thread

**Pull request merged**